frame/contracts/src/migration.rs

```diff
@@ -15,27 +15,34 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::{Config, Weight, CurrentSchedule, Pallet, Schedule};
-use frame_support::traits::{GetPalletVersion, PalletVersion, Get};
+use crate::{Config, Weight, Pallet, Schedule};
+use frame_support::{
+	storage::StorageValue,
+	traits::{GetPalletVersion, PalletVersion, Get},
+};
 
 pub fn migrate<T: Config>() -> Weight {
 	let mut weight: Weight = 0;
 
 	match <Pallet<T>>::storage_version() {
-		// Replace the schedule with the new default and increment its version.
 		Some(version) if version == PalletVersion::new(3, 0, 0) => {
-			weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
-			let _ = <CurrentSchedule<T>>::translate::<u32, _>(|version| {
-				version.map(|version| Schedule {
-						version: version.saturating_add(1),
-						// Default limits were not decreased. Therefore it is OK to overwrite
-						// the schedule with the new defaults.
-						.. Default::default()
-					})
-			});
+			weight = weight.saturating_add(T::DbWeight::get().writes(1));
+			v3_0_0::CurrentSchedule::<T>::kill();
 		}
 		_ => (),
 	}
 
 	weight
 }
+
+mod v3_0_0 {
+	use super::*;
+
+	struct Pallet<T: Config>(sp_std::marker::PhantomData<T>);
+
+	frame_support::decl_storage! {
+		trait Store for Pallet<T: Config> as Contracts {
+			pub CurrentSchedule: Schedule<T>;
```

Comment:
Just to be sure: this storage is using key at `twox128(Contracts)++...` whereas `CurrentSchedule` defined with pallet macro is using `twox128(name of contract pallet as defined in construct_runtime)++...`.

So if this upgrade is to be executed from a version of pallet-contract which defined its `CurrentSchedule` with pallet macro then this upgrade seems to make assumption about the name of pallet contract in construct_runtime.

Reply:
How would I go about this? I want to delete the storage item that was stored by a previous version of the contract.

Reply:
Previous version was this:
```rust
#[pallet::storage]
pub(crate) type CurrentSchedule<T: Config> = StorageValue<_, Schedule<T>, ValueQuery>;
```

I will go with the hardcoded pallet name. If you do funny things like changing pallet names during upgrades you need a custom migration. I will add a note for this to this PR and the changelog. It should be as easy as possible for the happy path. Otherwise we would force anyone to somehow supply this name somewhere (I don't know where).

Reply:
note that people have asked to be able to handle migration more precisely https://github.com/paritytech/substrate/issues/8713
But the implementation here is good to me